File: Rust/2022/18.rs
#![feature(test)]

use std::collections::VecDeque;

use itertools::Itertools;
use rustc_hash::FxHashSet;

type Input = Vec<Cube>;

#[derive(Clone, Hash, PartialEq, Eq)]
struct Cube(i16, i16, i16);

impl Cube {
    fn sides(&self) -> [CubeSide; 6] {
        [
            CubeSide(self.clone(), Dimension::X),
            CubeSide(self.clone(), Dimension::Y),
            CubeSide(self.clone(), Dimension::Z),
            CubeSide(Cube(self.0 - 1, self.1, self.2), Dimension::X),
            CubeSide(Cube(self.0, self.1 - 1, self.2), Dimension::Y),
            CubeSide(Cube(self.0, self.1, self.2 - 1), Dimension::Z),
        ]
    }

    fn neighbors(&self) -> [Self; 6] {
        [
            Cube(self.0 - 1, self.1, self.2),
            Cube(self.0 + 1, self.1, self.2),
            Cube(self.0, self.1 - 1, self.2),
            Cube(self.0, self.1 + 1, self.2),
            Cube(self.0, self.1, self.2 - 1),
            Cube(self.0, self.1, self.2 + 1),
        ]
    }
}

#[derive(Hash, PartialEq, Eq)]
struct CubeSide(Cube, Dimension);

#[derive(Hash, PartialEq, Eq)]
enum Dimension {
    X,
    Y,
    Z,
}

fn setup(input: &str) -> Input {
    input
        .trim()
        .lines()
        .map(|line| {
            let (x, y, z) = line
                .split(',')
                .map(|x| x.parse().unwrap())
                .collect_tuple()
                .unwrap();
            Cube(x, y, z)
        })
        .collect()
}

fn part1(input: &Input) -> usize {
    let mut sides = FxHashSet::with_capacity_and_hasher(input.len() * 6, Default::default());
    for cube in input {
        for side in cube.sides() {
            if sides.contains(&side) {
                sides.remove(&side);
            } else {
                sides.insert(side);
            }
        }
    }
    sides.len()
}

fn part2(input: &Input) -> usize {
    let mut out = 0;
    let (minx, maxx) = input.iter().map(|x| x.0).minmax().into_option().unwrap();
    let (miny, maxy) = input.iter().map(|x| x.1).minmax().into_option().unwrap();
    let (minz, maxz) = input.iter().map(|x| x.2).minmax().into_option().unwrap();
    let n = ((maxx - minx + 3) * (maxy - miny + 3) * (maxz - minz + 3)) as _;

    let cubes = input.iter().collect::<FxHashSet<_>>();
    let mut queue = VecDeque::with_capacity(n);
    queue.push_front(Cube(minx - 1, miny - 1, minz - 1));
    let mut visited = FxHashSet::with_capacity_and_hasher(n, Default::default());
    while let Some(p) = queue.pop_front() {
        if visited.contains(&p) {
            continue;
        }
        visited.insert(p.clone());

        for q in p.neighbors() {
            if !(minx - 1..=maxx + 1).contains(&q.0)
                || !(miny - 1..=maxy + 1).contains(&q.1)
                || !(minz - 1..=maxz + 1).contains(&q.2)
            {
                continue;
            }
            if cubes.contains(&q) {
                out += 1;
            } else if !visited.contains(&q) {
                queue.push_back(q);
            }
        }
    }

    out
}

aoc::main!(2022, 18, ex: 1);

File: Python/2022/18.py
from lib import *

input = read_input(2022, 18)


cubes = [(x, y, z) for x, y, z in map(ints, input.splitlines())]


out = 0
for i, x in enumerate(cubes):
    out += 6
    for y in cubes[:i]:
        if abs(x[0] - y[0]) + abs(x[1] - y[1]) + abs(x[2] - y[2]) == 1:
            out -= 2
print(out)


out = 0
(minx, miny, minz) = (maxx, maxy, maxz) = cubes[0]
for i, x in enumerate(cubes):
    out += 6
    for y in cubes[:i]:
        if abs(x[0] - y[0]) + abs(x[1] - y[1]) + abs(x[2] - y[2]) == 1:
            out -= 2
    minx = min(x[0], minx)
    miny = min(x[1], miny)
    minz = min(x[2], minz)
    maxx = max(x[0], maxx)
    maxy = max(x[1], maxy)
    maxz = max(x[2], maxz)

checked = set()
candidates = [
    (i, j, k)
    for x, y, z in cubes
    for i, j, k in [(x, y, z - 1), (x, y, z + 1), (x, y - 1, z), (x, y + 1, z), (x - 1, y, z), (x + 1, y, z)]
    if (i, j, k) not in cubes
]
while candidates:
    queue = [candidates.pop()]
    if queue[0] in checked:
        continue
    visited = set()
    while queue:
        x, y, z = queue.pop(0)
        if x not in range(minx, maxx + 1) or y not in range(miny, maxy + 1) or z not in range(minz, maxz + 1):
            break
        if (x, y, z) in visited:
            continue
        visited.add((x, y, z))
        for q in [(x, y, z - 1), (x, y, z + 1), (x, y - 1, z), (x, y + 1, z), (x - 1, y, z), (x + 1, y, z)]:
            if q not in cubes and q not in visited:
                queue.append(q)
    else:
        for x, y, z in visited:
            for q in [(x, y, z - 1), (x, y, z + 1), (x, y - 1, z), (x, y + 1, z), (x - 1, y, z), (x + 1, y, z)]:
                if q in cubes:
                    out -= 1

    checked.update(visited)

print(out)
